### PR TITLE
feat: Add top-k multilabel precision and recall metrics

### LINKED ISSUE #467: Top-K precision/recall multilabel metrics for ranking task
Following the discussion from https://github.com/pytorch/ignite/issues/466#issuecomment-478339986 it would be nice to have such metric in Ignite.

In the context of a multilabel task, compute a top-k precision/recall per label (treating all labels independently).

### PR DESCRIPTION
**Pull request by abdelmagid07**
Fixes #467 

## Description

Adds two new multilabel ranking metrics: `TopKMultilabelPrecision` and `TopKMultilabelRecall` to `ignite.metrics`.

For each sample, the `k` labels with the highest prediction scores are treated as positive predictions; precision and recall are then reported **per label** across the accumulated samples. This matches the metric described in the discussion (#466 -> #467 ), grounded in https://arxiv.org/pdf/1312.4894.pdf

## Notes:

-`y_pred` must be **float scores**.
- If `k` exceeds the number of labels, it is clamped to `num_labels`.
- Default k = 5

## Changes

- New module `ignite/metrics/top_k_multilabel_precision_recall.py`
- Export `TopKMultilabelPrecision` and `TopKMultilabelRecall` from `ignite.metrics`
- Add entries to `docs/source/metrics.rst`
- RST docstrings with `.. testcode::` / `.. testoutput::` examples
- Tests at `ignite/metrics/test_top_k_multilabel_precision_recall.py`

## Check list:

- [x] New tests are added (if a new feature is added)
- [x] New doc strings: description and/or example code are in RST format
- [x] Documentation is updated (if required)

### CODE REVIEW

File: ignite/metrics/top_k_multilabel_precision_recall.py [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+        self._k = k
+        super().__init__(
+            output_transform=output_transform,
+            average=average,
+            is_multilabel=True,
+            device=device,
+            skip_unrolling=skip_unrolling,
+        )
+
+    def _check_type(self, output: Sequence[torch.Tensor]) -> None:
```
Comment (aaishwarymishra) [inline ignite/metrics/top_k_multilabel_precision_recall.py:39]: The parent class implements this method, do we need to implement it?

File: ignite/metrics/top_k_multilabel_precision_recall.py [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+        k = min(self._k, num_labels)
+        topk_indices = torch.topk(y_pred, k, dim=1).indices
+        y_pred_topk = torch.zeros_like(y_pred)
+        y_pred_topk.scatter_(1, topk_indices, 1.0)
+
+        y_pred_topk = y_pred_topk.to(dtype=self._double_dtype, device=self._device)
+        y = y.to(dtype=self._double_dtype, device=self._device)
+        correct = y * y_pred_topk
+
+        return y_pred_topk, y, correct
```
Comment (aaishwarymishra) [inline ignite/metrics/top_k_multilabel_precision_recall.py:76]: I think other classification metrics like `Accuracy` follow `(N,S,C)` batch, sequence, classes I think this metric should be consistent with other metrics.

File: ignite/metrics/top_k_multilabel_precision_recall.py
Comment (aaishwarymishra) [inline ignite/metrics/top_k_multilabel_precision_recall.py:1]: Isn't essentially both classes doing the same thing why we need 2 separate classes?

File: ignite/metrics/top_k_multilabel_precision_recall.py (hunk 1 above)
Comment (abdelmagid07) [reply ignite/metrics/top_k_multilabel_precision_recall.py:39]: I think we do because the parent's method requires y_pred to be binary. These metrics take raw prediction scores in float form, such as 0.8, 0.6, etc., and select the top-k values, so the original check can't be reused; the parent method would end up rejecting valid inputs. The override validates float y_pred instead, while keeping the same other checks. I can add comments in the code addressing this.

File: ignite/metrics/top_k_multilabel_precision_recall.py (hunk 2 above)
Comment (abdelmagid07) [reply ignite/metrics/top_k_multilabel_precision_recall.py:76]: I aimed to mirror the existing Precision and Recall, which use the (N, C, ...) shape,
Like in precision.py (line 78 - 82):

        elif self._type == "multilabel":
            # if y, y_pred shape is (N, C, ...) -> (N * ..., C)
            num_labels = y_pred.size(1)
            y_pred = torch.transpose(y_pred, 1, -1).reshape(-1, num_labels)
            y = torch.transpose(y, 1, -1).reshape(-1, num_labels)

I see the (N, S, C) shape in get_sequence_transform in accuracy.py, but this seems like a different case than multilabel scoring. The tests cover (N, C), (N, C, L), and (N, C, H, W) inputs and pass, so the extra dimensions are handled the same way as the other multilabel metrics. Let me know your thoughts here.

File: ignite/metrics/top_k_multilabel_precision_recall.py
Comment (abdelmagid07) [reply ignite/metrics/top_k_multilabel_precision_recall.py:1]: My goal there was to mirror the existing API as much as possible. From my understanding, we already ship Precision and Recall as two separate classes over the shared parent _BasePrecisionRecall, with update methods that differ in the denominator they use (predicted vs actual positives). I tried to mirror that by doing the same over _BaseTopKMultilabelPrecisionRecall. Essentially, we would have TopKMultilabelPrecision/TopKMultilabelRecall reading parallel to the existing Precision/Recall split. Though I'm happy to merge them into a single class if that fits the codebase conventions better.

**Comment by aaishwarymishra** [comment]
@abdelmagid07 To be honest, looking at the code I am not sure how these metrics work, the compute of the precision and recall will differ right, but there is no compute implementation, wont the results of both metrics be same when we do .compute() , even when we have 2 completely different metrics precision and recall.

**Comment by vfdev-5** [comment]
@abdelmagid07 @aaishwarymishra I would say we would better to go with https://github.com/pytorch/ignite/issues/3568 approach instead of creating TopK version for each metric. Let's close this PR as a no-go.
